src/Flicking.ts

```diff
@@ -531,6 +546,90 @@ class Flicking extends Component {
+   * @param {number[]} [diffInfo.removed] - Index array of DOM elements removed from previous element list.<ko>이전 리스트에서 제거된 DOM 요소들의 인덱스 배열.</ko>
+   */
+  public sync(diffInfo: {
+    list: HTMLElement[],
+    maintained: number[][],
+    added: number[],
+    removed: number[],
+  }) {
+    const { list, maintained, added, removed } = diffInfo;
+
+    const viewport = this.viewport;
+    const panelManager = viewport.panelManager;
+    const indexRange = panelManager.getRange();
+    const isCircular = this.options.circular;
+
+    // Make sure that new "list" should include cloned elements
+    const newOriginalPanelCount = (list.length / (panelManager.getCloneCount() + 1)) >> 0; // Make sure it's integer. Same with Math.floor, but faster
+    const newCloneCount = ((list.length / newOriginalPanelCount) >> 0) - 1;
+
+    const prevOriginalPanels = panelManager.originalPanels();
+    const prevClonedPanels = panelManager.clonedPanels();
+
+    const newOriginalElements = list.slice(0, newOriginalPanelCount);
+    const newClonedElements = list.slice(newOriginalPanelCount);
+
+    const newPanels: Panel[] = [];
+    const newClones: Panel[][] = counter(newCloneCount).map(() => []);
+
+    // For maintained panels after external rendering, they should be maintained in newPanels.
+    const originalMaintained = maintained.filter(([beforeIdx, afterIdx]) => beforeIdx <= indexRange.max);
+    // For newly added panels after external rendering, they will be added with their elements.
+    const originalAdded = added.filter(index => index < newOriginalPanelCount);
+
+    originalMaintained.forEach(([beforeIdx, afterIdx]) => {
+      newPanels[afterIdx] = prevOriginalPanels[beforeIdx];
+      newPanels[afterIdx].setIndex(afterIdx);
+    });
+
+    originalAdded.forEach(addIndex => {
+      newPanels[addIndex] = new Panel(newOriginalElements[addIndex], addIndex, viewport);
+    });
+
+    if (isCircular) {
+      counter(newCloneCount).forEach(groupIndex => {
+        const cloneGroupOffset = newOriginalPanelCount * groupIndex;
+        const prevCloneGroup = prevClonedPanels[groupIndex];
+        const newCloneGroup = newClones[groupIndex];
+
+        originalMaintained.forEach(([beforeIdx, afterIdx]) => {
+          newCloneGroup[afterIdx] = prevCloneGroup
+            ? prevCloneGroup[beforeIdx]
+            : newPanels[afterIdx].cloneExternal(groupIndex, newClonedElements[cloneGroupOffset + afterIdx]);
+        });
+
+        originalAdded.forEach(addIndex => {
+          const newPanel = newPanels[addIndex];
+
+          newCloneGroup[addIndex] = newPanel.cloneExternal(groupIndex, newClonedElements[cloneGroupOffset + addIndex]);
+        });
+      });
+    }
+
+    // Replace current info of panels this holds
+    added.forEach(index => { viewport.updateCheckedIndexes({ min: index, max: index }); });
+    removed.forEach(index => { viewport.updateCheckedIndexes({ min: index - 1, max: index + 1 }); });
```

Comment:
@WoodNeck Is order important? eg. `added -> removed` or `removed ->added`

Reply:
I think not. Only order of `maintained` is important.